### PR TITLE
adds Public Permalinks

### PR DESCRIPTION
**Pull request by jeremybmerrill**
I included tests too!

Includes the routing work in Rustland, along with the React stuff.

### CODE REVIEW

File: backend/server/public/src/components/addetail.jsx [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import { withRouter, Link } from "react-router-dom";
+import { getOneAd } from "../actions.js";
+
+export class AdDetailUnconnected extends React.Component {
+  componentDidMount() {
+    // if this.props.ad, we're good, we got the ad prop from <Frontend> or it's already been fetched.
+    // if !this.props.ad && !this.props.getOneAd, then something has gone wrong.
+    if (
+      !(this.props.ad && this.props.ad.id) &&
+      this.props.getOneAd &&
```
Comment (thejefflarson) [inline backend/server/public/src/components/addetail.jsx]: What's `getOneAd`? This seems really complicated.

File: backend/server/public/src/components/addetail.jsx [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      this.props.match
+    ) {
+      let ad_id = null;
```
Comment (thejefflarson) [inline backend/server/public/src/components/addetail.jsx]: combine these two lines and prefer `const`

File: backend/server/public/src/components/frontend.jsx [hunk 3]
```diff
@@ -11,49 +13,64 @@ import { t } from "i18n.js";
...
-      ) : (
-        <p className="no_ads">No ads found for {search}.</p>
-      )}
-      <div id="ads">{ads.map(ad => <Ad ad={ad} key={ad.id} />)}</div>
-      {ads.length > 0 ? <Pagination /> : ""}
-    </div>
+    <Switch>
+      <Route exact path="/facebook-ads">
+        {/* confusingly, despite being `exact`, this also matches /facebook-ads/, with the trailing slash */}
+        <div>
```
Comment (thejefflarson) [inline backend/server/public/src/components/frontend.jsx]: we should probably make this a component, no?

File: backend/server/src/server.rs [hunk 4]
```diff
@@ -141,7 +141,9 @@ impl Service for AdServer {
                 let restful = RegexSet::new(&[ // rudimentary routing. ORDER MATTERS. And we're using the index of these as the key for match below.
                     r"^/facebook-ads/ads/?(\d+)?$",
                     r"^/facebook-ads/admin/?(.*)?$",
-                    r"^/facebook-ads(/?)$" // TODO: this will never serve a 404 and will always render the index...
+                    r"^/facebook-ads(/?)$",
+                    r"^/facebook-ads/ad/?(\d+)?$"
+
```
Comment (thejefflarson) [inline backend/server/src/server.rs]: kill this empty line

File: backend/server/src/server.rs [hunk 5]
```diff
@@ -141,7 +141,9 @@ impl Service for AdServer {
...
                 ]).unwrap();
                 let restful_collection_element_regex = Regex::new(r"^/facebook-ads/(?:[^/]+)/?(\d+)$").unwrap(); // generic restful routing regex for distinguishing subroutes at the collection and those at a specific element.
```
Comment (thejefflarson) [inline backend/server/src/server.rs]: kill this line two.

File: backend/server/public/src/components/addetail.jsx (hunk 1 above)
Comment (jeremybmerrill) [reply backend/server/public/src/components/addetail.jsx]: It's the function that goes to the Rust API to get the data for the permalinked ad.

**Comment by jeremybmerrill** [comment]
good call; I adjusted that to use propTypes. Also added tests for AdList (the factored-out bit of Frontend.js).

**Pull request closed without merging**